### PR TITLE
Proper World Info logic checks

### PR DESCRIPTION
**Pull request by Bronya-Rand**
This PR does the following;
 
- Reverts PR #1356 as the fix made by this PR is logically flawed.
> The flawed logic comes from a polar opposite check for the NOT condition and didn't really differ from the AND check itself.
- Rewrites PR #1356's logic using AND ONE check as a base and doing the polar opposite of it (aka using `!`)
- Combines the AND ONE/NOT ONE checks in 1 If statement for compactness.
- Adjusted debugger text for better debugging purposes.

## New Changes
- Added NOT ALL (requested by Cohee)
- Renamed AND to AND ONE and NOT to NOT ONE (Cohee)

### CODE REVIEW

**Comment by nonenothingnada** [comment]
Some important notes on this change:

1. There are now 3/4 possible interpretations of the primarykey/secondarykey + AND/NOT operators and at least 1 of them is currently intuitively inconsistent with the previous state of the interface (or requires a pretty big logical leap, see point 4).


2. It would be more clear to rename these as follows:
    - AND_ONE => AND_ANY
    - (AND_ALL) => AND_ALL (to be implemented)
    - NOT_ONE => AND_NOT_ANY
    - NOT_ALL => AND_NOT_ALL


> The flawed logic comes from a polar opposite check for the NOT condition and didn't really differ from the AND check itself.
3. I'm not sure what this word salad is meant to mean, see points 4 and 5.


4. In its current state we have two fields (primary and secondary keys) and a drop down for an operator. Previously this drop down only contained AND and NOT. Given that the original intent wasn't really specified we need to clarify what these ought to (have) mean(t). The primary and secondary keys accept comma delimited lists of terms. When filling out the primary keys only the terms are treated as being separated by an implicit OR. As no other information is present to imply otherwise it should be assumed the same is true for the secondary keys field. To achieve the current implementation of NOT_ALL using the previous NOT operator we need to instead assume the terms are separated by an implicit AND. Let's consider the previously working case of an AND operator with one primary key and several secondary keys, this is the equivalent of:


5. Using the following symbols: (v = OR, & = AND, ~ = NOT)
> P & (Q v R)
- This means we activate if we have P and any of (Q, R). This makes sense and is consistent with what was implemented. Let's take a look at the NOT operator now. One thing to consider though is that NOT is unary so there should also be an implicit AND in front of it.
> P & ~(Q v R)
- This is what should be logically implied by a WI entry using the NOT operator and several secondary keys. We could fill out a truth table to determine what the outcome should be for all cases, or we can just apply https://en.wikipedia.org/wiki/De_Morgan's_laws. This is what we get:
> P & (~Q & ~R)
- If we have P we should activate if we have neither Q nor R. This is consistent with the current NOT_ONE operator, though for the reasons stated above and as 3/4 of these have been implemented it would be helpful to rename this to AND_NOT_ANY.
> P & (~Q v ~R)
- This would be an unexpected outcome of the NOT operator as it's an improper application of De Morgan's, clarifying it to NOT_ALL helps, though to be more formally correct it might further be clarified to AND_NOT_ALL. In order for this to be the expected outcome we'd need to assume that, for the NOT case, the terms were separated by AND as follows:
> P & ~(Q & R)
- Which becomes the previously stated P & (~Q v ~R) meaning if we have P we should activate if we don't have both Q and R.


6. As someone who uses software it's also important to consider what the likely use cases of this functionality are. Let's assume a primary key of COAT and secondary keys of RAIN and WINTER:
    - NULL operator (only primary keys used): Very clear, enrich the context with information about COAT, whatever that means in the current scenario.
    - AND_ANY (AND_ONE): Also very clear, enrich the context with information related to the conjunction of the primary+secondary keys - a RAIN COAT or a WINTER COAT (though these terms don't necessarily need to appear side-by-side) when we don't care about a WIND COAT or a SUMMER COAT.
    - AND_ALL: Less clear but still potentially useful. In this case we'd only activate if we see both RAIN and WINTER in conjunction with COAT. Maybe this only matters when it's raining in winter as opposed to snowing?
    - AND_NOT_ANY (NOT_ONE): Niche but can be very useful. In this case we only care about the COAT if we see neither RAIN nor WINTER. Why would we do this? Why is someone wearing a coat when it's neither cold nor raining? This might be a good indicator of a novel situation or that things are going off the rails and we can use it to enrich the context or try to course correct the scenario.
    -  AND_NOT_ALL (NOT_ALL): This one I have a more difficult time finding good use cases for. We care about the coat so long as it's not both RAIN'ing and WINTER? Yes this probably means that it's snowing, or maybe it's windy, or maybe something else. It's not like this couldn't be useful, it's just less immediately clear how one might use it. It's also not logically consistent without assuming that swapping to the NOT operator also swaps the implicit OR's to AND's separating the terms.


tl;dr: There should be 2 operators (AND and NOT with proper implementations) or 4. If there are 4 there needs to be additional context present to properly specify the logical operation (hence the renaming). Even with renaming AND_NOT_ANY and AND_NOT_ALL are a little ambiguous, but NOT_ONE and NOT_ALL seem about the same to worse (and are lacking the implicit AND). The comment in the PR is incorrect and re-implements a bug (though thankfully the third operator re-enables the correct functionality). For simplified reasoning assume primarykey can be blank and reference: https://en.wikipedia.org/wiki/De_Morgan's_laws#Text_searching

**Comment by Cohee1207** [comment]
@nonenothingnada The point of splitting these to NOT ANY and NOT ALL was that the sudden change in logic was repeatedly found to be confusing and broke existing World Info books in unexpected ways. That way you have access to both new and old NOT checks in more explicit ways. The naming part and the new logic can be considered for further improvements.

**Comment by nonenothingnada** [comment]
>broke existing World Info books in unexpected ways
- The previous logic only considered the first entry in secondarykeys. When scanning it would consider the first primary/secondary key combination and bail out or immediately activate (and potentially subsequently activate more times, though this doesn't seem to have any effect). It should be expected that WI books that used NOT behave differently after this was corrected because the behavior was caused by both a bug and incorrect (or at least counter-intuitive) logic. Also, for entries with a single secondarykey the behavior should be identical between both NOTs.

>sudden change in logic was repeatedly found to be confusing
>That way you have access to both new and old NOT checks in more explicit ways
- Confusing from the previous implementation? That's reasonable. I'm not sure where the "flawed logic" word salad is coming from. Still, implementing 3 / 4 operators (as flipping the tacit OR to an AND in the secondary keys is now considered valid) is more confusing.

**Comment by Cohee1207** [comment]
> The previous logic only considered the first entry in secondarykeys

That was an issue, but considering ALL keys instead of just ONE was not expected here.

As for other questions, open a discussion thread if you find this concerning or anything. A closed PR is not a proper communication channel.
